submission/distributor.go

```diff
@@ -0,0 +1,130 @@
+	"github.com/google/certificate-transparency-go/trillian/ctfe"
+)
+
+const (
+	// RootsRefreshInterval is interval between consecutive get-roots calls.
+	rootsRefreshInterval = time.Hour * 24
+)
+
+// Distributor operates policy-based submission across Logs.
+type Distributor struct {
+	ll *loglist.LogList
+
+	mu sync.RWMutex
+
+	// helper structs produced out of ll during init.
+	logClients map[string]client.AddLogClient
+	logRoots   loglist.LogRoots
+	rootPool   *ctfe.PEMCertPool
+
+	rootsRefreshTicker *time.Ticker
+
+	policy ctpolicy.CTPolicy
+}
+
+// Run starts regular roots updates.
+func (d *Distributor) Run(ctx context.Context) {
+	if d.rootsRefreshTicker != nil {
+		return
+	}
+	d.rootsRefreshTicker = time.NewTicker(rootsRefreshInterval)
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-d.rootsRefreshTicker.C:
+				d.refreshRoots(ctx)
+			}
+		}
+	}()
+}
+
+// refreshRoots requests roots from Logs and updates local copy if necessary.
+func (d *Distributor) refreshRoots(ctx context.Context) {
+	// TODO(Mercurrent) add implementation.
+}
+
+// SubmitToLog is races-Submitter interface.
+func (d *Distributor) SubmitToLog(ctx context.Context, logURL string, chain []ct.ASN1Cert) (*ct.SignedCertificateTimestamp, error) {
+	// TODO(Mercurrent) add implementation.
+	return nil, nil
+}
+
+// AddPreChain runs add-pre-chain calls across subset of logs according to
+// Distributor's policy. May emit both SCTs array and error when SCTs
+// collected do not satisfy the policy.
+func (d *Distributor) AddPreChain(ctx context.Context, rawChain [][]byte) ([]*AssignedSCT, error) {
+	// TODO(Mercurrent) add implementation.
+	return []*AssignedSCT{}, nil
+}
+
+// LogClientBuilder builds client-interface instance for a given Log.
+type LogClientBuilder func(*loglist.Log) (client.AddLogClient, error)
+
+// BuildLogClient is default (non-mock) LogClientBuilder.
+func buildLogClient(log *loglist.Log) (client.AddLogClient, error) {
+	url, err := url.Parse(log.URL)
+	if err != nil {
+		return nil, err
+	}
+	if url.Scheme == "" {
+		url.Scheme = "https"
+	}
+	hc := &http.Client{Timeout: time.Second * 10}
+	return client.New(url.String(), hc, jsonclient.Options{PublicKeyDER: log.Key})
+}
+
+// NewDistributor creates and inits a Distributor instance.
+// To get active, Distributor is expected to call its Run() method.
+func NewDistributor(ll *loglist.LogList, plc ctpolicy.CTPolicy, lcBuilder LogClientBuilder) (*Distributor, error) {
+	var d Distributor
+	active := ll.ActiveLogs()
+	d.ll = &active
+	d.policy = plc
+	d.logClients = make(map[string]client.AddLogClient)
+	d.logRoots = make(loglist.LogRoots)
+
+	// Build clients for each of the Logs.
+	for _, log := range ll.Logs {
+		lc, err := lcBuilder(&log)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create log client for %s: %v", log.URL, err)
+		}
+		d.logClients[log.URL] = lc
+	}
+	// Collect Log-roots.
+	go func() {
+		d.refreshRoots(context.Background())
```

Comment:
At the point that NewDistributor() returns, you cannot deterministically say that Distributor is initialized (if fetching the roots is part of initialization) since you do it asynchronously. It should either be done synchronously, if it's a necessary part of initialization, or deferred to when the user calls `Run()` if it isn't necessary before then.

Reply:
Done